README.md
# [![ClassIn-Mover icon 32x32](ClassIn_Mover_32.png) ClassIn-Mover ![GitHub all releases](https://img.shields.io/github/downloads/CarlGao4/Classin-mover/total)](https://classin-mover.pages.dev/)

[简体中文](https://classin-mover.pages.dev/zh-cn)

[Download latest version](https://classin-mover.pages.dev/)

A program to move `ClassIn` classroom window in order to exit from focused learning mode.

Supported `ClassIn` version: `3.0.2.130` to `3.0.5.1`, as well as `3.0.7.x` `4.x`. (`ClassIn` prior to `3.0.2.130` does not have focused learning mode at all)

Maybe the program will lose efficacy in the future versions.

## Components

### Classic

The program remained the old style of v1.0.0, except the modern UI by Tk. 

#### Usage

Run this program before entering the classroom, then get into the classroom as normal.

After that, the program will automatically detect the classroom window, and make it unable to occupy the whole screen.

If working well, the program should output one line of log each second, showing the current working status.

### Enhanced

The program will be started on the minimized status, which only a transparent icon is displayed. You can click it to open the main window, which can move, resize, and set z-order (or topmost) of your selected ClassIn window. 

Auto patch is enabled as default, which will automatically resize the window into normal state like other applications once a new ClassIn window has been detected. You can disable this feature by right-clicking the minimized icon. The button `Auto Patch` on the main window will not enable this feature, but just patch the selected window.

## Running the codes

As you may have noticed, the latest version uses Python, which means you can run the codes yourself. The dependencies are `Pillow` and `psutil`. After you have installed these two packages using `pip` or `conda`, you can directly run the program with a Python interpreter. 

**However, it is not recommended to run or modify the codes if you don't have coding experience. Use the releases instead.**

## Known issues

- **This program calls native API and thus can only work on Windows.**

## Donate

![WeChat](img/wechat.png)

![AliPay](img/alipay.jpg)


history.md
## 1.0.0

> Date: 2020/6/30

- The first stable release of `ClassIn-Mover`.

## 1.0.1

> Date: 2020/7/2

- Optimize window searching mechanism and reduce log delay.

## 1.0.2

> Date: 2020/7/3

- Optimized delay mechanism.

## 1.0.3

> Date: 2020/7/6

- Increased log level.

## 1.0.4

> Date: 2021/1/15

- Added auto get administrator's permission when opening the program if you didn't run as administrator. This mode is added due to sometimes it doesn't patch the window even the program outputs "processed". If you don't give the authority, the program will run normally and output a warning.

- Warnings will be printed in yellow and errors will be printed in red.

## 1.0.5

> Date: 2022/6/13

- Optimise memory.

---

## 2.0.0 ![Downloads](https://img.shields.io/github/downloads/CarlGao4/ClassIn-Mover/v2.0.0/total)

> Date: 2022/11/24

- Brand new version v2.0.0

- New modern ClassIn Mover (Compared with classIn Mover Classic)

- Graphic User Interface

- Manually minimize, maximize, topmost, no-topmost the ClassIn window

- Drag to move and resize ClassIn window

- Minimize to floating icon

- Stop supporting ClassIn Mover Classic

## 2.0.1 ![Downloads](https://img.shields.io/github/downloads/CarlGao4/ClassIn-Mover/v2.0.1/total)

> Date: 2022/11/25

- Add auto check updates. 

## 2.0.2 ![Downloads](https://img.shields.io/github/downloads/CarlGao4/ClassIn-Mover/v2.0.2/total)

> Date: 2022/11/26

- Fix bug: Window is full screen most of time after patching

- Add "Patch all" to right-click menu

- Auto patch window will now delay 8.5s when a new window detected, to ensure that the classroom has successfully entered. 

## 2.1.0 ![Downloads](https://img.shields.io/github/downloads/CarlGao4/ClassIn-Mover/v2.1.0/total)

> Date: 2022/11/28

- Chinese support

- Add button to open our homepage

- Fix bug: May receive message box even the main window has been closed

## 2.1.0.1 ![Downloads](https://img.shields.io/github/downloads/CarlGao4/ClassIn-Mover/v2.1.0.1/total)

> Date: 2022/11/28

- Fix bug: Chinese translation cause can't move or resize window

## 2.1.0.2 ![Downloads](https://img.shields.io/github/downloads/CarlGao4/ClassIn-Mover/v2.1.0.2/total)

> Date: 2022/12/2

- Fix bug: program may stop working after long time

- Fix bug: ClassIn window size can be changed into negative number which will cause ClassIn stop responding

- Add "Usage" button

## 2.1.1 ![Downloads](https://img.shields.io/github/downloads/CarlGao4/ClassIn-Mover/v2.1.1/total)

> Date: 2022/12/5

- Fix grammar and translation

- The main window will appear beside the floating window

- Add alternative update link

- Optimize codes

## 2.1.2 ![Downloads](https://img.shields.io/github/downloads/CarlGao4/ClassIn-Mover/v2.1.2/total)

> Date: 2022/12/12

- Fix bug: window will still be patched if "Auto patch new window" is disabled after a new classroom window is opened and before the 8.5s delay expires

- Add "Blogs" button to view our latest articles

- The settings will be preserved after the program exits

## 2.1.3 ![Downloads](https://img.shields.io/github/downloads/CarlGao4/ClassIn-Mover/v2.1.3/total)

> Date: 2022/12/14

- Posts on the main window

- See latest blog list

- Arguments for debugging

## 2.2.0 ![Downloads](https://img.shields.io/github/downloads/CarlGao4/ClassIn-Mover/v2.2.0/total)

> Date: 2022/12/26

- Remove classroom watermark

- On-stage notification (available choices: Message box, Switching to current window, Closing classroom window, Closing ClassIn) (Maybe unavailable on some systems due to system settings, still under investigation)

- Auto start with logging on

- Manually restart as administrator

- Set floating icon transparency

- Optimize keep topmost waiting mechanism

- Increase log level

- Use `msi` as installer

## 2.2.0.1 ![Downloads](https://img.shields.io/github/downloads/CarlGao4/ClassIn-Mover/v2.2.0.1/total)

> Date: 2022/12/26

- Fix bug: May not be notified on some systems

shlex4all.py
"""
## shlex for all

The standard lib `shlex` provides parsing only for Unix, this library is
intended to provide such parsing for all platforms. You can use `join`,
`split` and `quote` function.

On Windows, the parsing follows documents on https://docs.python.org/3.11/library/subprocess.html#converting-an-argument-sequence-to-a-string-on-windows and https://learn.microsoft.com/en-us/windows/win32/api/shellapi/nf-shellapi-commandlinetoargvw#remarks

On other platforms, the functions are same as `shlex`. 
"""

import re
import sys
from typing import Iterable

__all__ = ["join", "quote"]

if sys.platform != "win32":
    import shlex

    join = shlex.join
    split = shlex.split
    quote = shlex.quote

else:

    def quote(s: str):
        if type(s) != str:
            raise TypeError("s must be `str`")
        if " " not in s and "\t" not in s and '"' not in s:
            return s
        ret = s
        if '"' in ret:
            ret = ret.replace('"', '\\"')
        if " " in ret or "\t" in ret:
            ret = '"' + ret + '"'
        ret = re.sub('(\\\\+)\\\\"', '\\1\\1\\"', ret)
        return ret


    def join(s: Iterable):
        return " ".join(quote(i) for i in s)


ClassIn_Mover.cpp
/*
    ClassIn Mover - A program to move ClassIn classroom window in order to
    exit from focused learning mode.
    Copyright (C) 2020-2021  Weiqi Gao, Jize Guo

    This program is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, version 3 of the License.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program.  If not, see <https://www.gnu.org/licenses/>.
*/

#include <cstdio>
#include <cstring>
#include <ctime>
#include <windows.h>
#include <tlhelp32.h>
#include <Psapi.h>
#pragma comment (lib, "Psapi.lib")
#pragma warning(disable:4996)

HWND ClassroomHwnd;
bool FoundWindow;
TCHAR title[1024];

BOOL CALLBACK RefreshWindow(HWND hwnd, LPARAM Title)
{
    PROCESSENTRY32 pe32;
    pe32.dwSize = sizeof pe32;
    time_t tmp = time(0);
    tm* now = localtime(&tmp);
    TCHAR caption[1024];
    memset(caption, 0, 1024);
    GetWindowText(hwnd, caption, 1024);
    DWORD pid;
    if (lstrlen(caption) && wcsstr(caption, L"Classroom_") != NULL && !FoundWindow)
    {
        char FormattedTime[9];
        strftime(FormattedTime, 9, "%H:%M:%S", now);
        GetWindowThreadProcessId(hwnd, &pid);
        HANDLE hProcessSnap = ::CreateToolhelp32Snapshot(TH32CS_SNAPPROCESS, 0);
        if (hProcessSnap == INVALID_HANDLE_VALUE)
        {
            SetConsoleTextAttribute(GetStdHandle(STD_OUTPUT_HANDLE), FOREGROUND_INTENSITY | FOREGROUND_RED);
            printf("[%s] Failed to excute CreateToolhelp32Snapshot\n", FormattedTime);
            SetConsoleTextAttribute(GetStdHandle(STD_OUTPUT_HANDLE), FOREGROUND_INTENSITY | FOREGROUND_RED | FOREGROUND_GREEN | FOREGROUND_BLUE);
            return TRUE;
        }
        BOOL bMore = ::Process32First(hProcessSnap, &pe32);
        bool flag = false;
        while (bMore)
        {
            if (DWORD(pe32.th32ProcessID) == pid)
            {
                flag = true;
                break;
            }
            bMore = ::Process32Next(hProcessSnap, &pe32);
        }
        TCHAR path[MAX_PATH];
        wcscpy(path, pe32.szExeFile);
        wcslwr(path);
        if (!flag || wcsstr(path, L"classin") == NULL)
        {
            CloseHandle(hProcessSnap);
            return TRUE;
        }
        wcscpy(title, caption);
        ClassroomHwnd = hwnd;
        FoundWindow = true;
        CloseHandle(hProcessSnap);
    }
    return TRUE;
}

bool IsProcessRunAsAdmin()
{
    SID_IDENTIFIER_AUTHORITY NtAuthority = SECURITY_NT_AUTHORITY;
    PSID AdministratorsGroup;
    BOOL b = AllocateAndInitializeSid(
        &NtAuthority,
        2,
        SECURITY_BUILTIN_DOMAIN_RID,
        DOMAIN_ALIAS_RID_ADMINS,
        0, 0, 0, 0, 0, 0,
        &AdministratorsGroup);
    if (b)
    {
        CheckTokenMembership(NULL, AdministratorsGroup, &b);
        FreeSid(AdministratorsGroup);
    }
    return b == TRUE;
}

long long CurrentTime() {
    SYSTEMTIME st;
    GetLocalTime(&st);
    time_t tmp = time(0);
    return tmp + st.wMilliseconds;
}

void GetWindowCmd(UINT showCmd, char* ReturnValue)
{
    switch (showCmd) {
    case (UINT)SW_MINIMIZE: strcpy(ReturnValue, "Minimized"); break;
    case (UINT)SW_SHOWMINIMIZED: strcpy(ReturnValue, "Minimized"); break;
    case (UINT)SW_MAXIMIZE: strcpy(ReturnValue, "Maximized"); break;
    case (UINT)SW_NORMAL: strcpy(ReturnValue, "Normal"); break;
    default: sprintf(ReturnValue, "%d", showCmd); break;
    }
}

#define SleepUntilNext(ms) Sleep((ms) - CurrentTime() % (ms))

int main()
{
    SetConsoleTextAttribute(GetStdHandle(STD_OUTPUT_HANDLE), FOREGROUND_INTENSITY | FOREGROUND_RED | FOREGROUND_GREEN | FOREGROUND_BLUE);
    puts("ClassIn Mover v1.0.5");
    puts("Copyright (C) 2020-2022  Weiqi Gao, Jize Guo");
    puts("Visit https://github.com/CarlGao4/ClassIn-Mover for more information.\n");
    TCHAR szPath[MAX_PATH];
    ZeroMemory(szPath, MAX_PATH);
    ::GetModuleFileName(NULL, szPath, MAX_PATH);
    HINSTANCE res;
    if (!IsProcessRunAsAdmin())
    {
        res = ShellExecute(NULL, L"runas", szPath, NULL, NULL, SW_SHOW);
        if ((int)res > 32)
        {
            return 0;
        }
        SetConsoleTextAttribute(GetStdHandle(STD_OUTPUT_HANDLE), FOREGROUND_INTENSITY | FOREGROUND_RED | FOREGROUND_GREEN);
        printf("Warning : You do not have administrator's permission running this program. \n\n");
        SetConsoleTextAttribute(GetStdHandle(STD_OUTPUT_HANDLE), FOREGROUND_INTENSITY | FOREGROUND_RED | FOREGROUND_GREEN | FOREGROUND_BLUE);
    }
    RECT rect;
    HWND hwnd = GetDesktopWindow();
    GetWindowRect(hwnd, &rect);
    WINDOWPLACEMENT wp;
    wp.length = sizeof(WINDOWPLACEMENT);
    int processes;
    while (true)
    {
        FoundWindow = false;
        EnumWindows(RefreshWindow, NULL);
        time_t tmp = time(0);
        tm* now = localtime(&tmp);
        char FormattedTime[9];
        strftime(FormattedTime, 9, "%H:%M:%S", now);
        if (!FoundWindow)
        {
            printf("[%s] Cannot find classroom window\n", FormattedTime);
            SleepUntilNext(1000);
            continue;
        }
        processes = 0;
        GetWindowPlacement(ClassroomHwnd, &wp);
        UINT firstStatus = wp.showCmd;
        while (CurrentTime() % 1000 <= 800)
        {
            GetWindowRect(ClassroomHwnd, &rect);
            SetWindowPos(ClassroomHwnd, HWND_NOTOPMOST, rect.left, rect.top,
                rect.right - rect.left, rect.bottom - rect.top, SWP_NOSENDCHANGING);
            GetWindowPlacement(ClassroomHwnd, &wp);
            if (wp.showCmd != SW_MAXIMIZE && wp.showCmd != SW_MINIMIZE && wp.showCmd != SW_SHOWMINIMIZED)
            {
                ShowWindow(ClassroomHwnd, SW_MINIMIZE);
                ShowWindow(ClassroomHwnd, SW_MAXIMIZE);
            }
            processes++;
            Sleep(50);
        }
        char StatusString[10];
        GetWindowCmd(firstStatus, StatusString);
        printf("[%s] Classroom window is found and processed %d times:\n           HWND=%I64d title=%ls status=%s\n",
            FormattedTime, processes, (long long)hwnd, title, StatusString);
        SleepUntilNext(1000);
    }
}


sources/shared.py
# ClassIn Mover - A program to move ClassIn classroom window in order to
# exit from focused learning mode.

# Copyright (C) 2020-2022  Weiqi Gao, Jize Guo, Yiming Geng

# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, version 3 of the License.

# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.

# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <https://www.gnu.org/licenses/>.

import ctypes
import json
import logging
import pathlib
import traceback

_threads_total = 0


def ShowThread(func):
    def ThreadTrace(*args, **kw):
        global _threads_total
        _threads_total += 1
        tid = _threads_total
        logging.debug("Thread %d (%s) started" % (tid, func.__name__))
        res = func(*args, **kw)
        logging.debug("Thread %d (%s) stopped" % (tid, func.__name__))
        return res

    return ThreadTrace


SettingsFile = pathlib.Path.home() / "AppData" / "Local" / "ClassIn-Mover" / "settings.json"
if SettingsFile.exists():
    try:
        with open(str(SettingsFile), mode="rt", encoding="utf8") as f:
            settings = json.loads(f.read())
        if type(settings) != dict:
            raise TypeError
    except:
        settings = {}
else:
    settings = {}


def SetSetting(attr, value):
    global settings, SettingsFile
    logging.debug('(%s) Set setting "%s" to %s' % (traceback.extract_stack()[-2].name, attr, str(value)))
    settings[attr] = value
    with open(str(SettingsFile), mode="wt", encoding="utf8") as f:
        f.write(json.dumps(settings, separators=(",", ":")))


def GetSetting(attr, default=None, autoset=True):
    global settings
    if attr in settings:
        return settings[attr]
    else:
        if autoset:
            SetSetting(attr, default)
        return default


def SwitchWindow(hwnd):
    dwForeID = ctypes.windll.user32.GetWindowThreadProcessId(ctypes.windll.user32.GetForegroundWindow(), 0)
    dwCurID = ctypes.windll.kernel32.GetCurrentThreadId()
    ctypes.windll.user32.AttachThreadInput(dwCurID, dwForeID, 1)
    ctypes.windll.user32.ShowWindow(hwnd, 5)
    ctypes.windll.user32.SetWindowPos(hwnd, 0, 0, 0, 0, 0, 7)
    ctypes.windll.user32.SetForegroundWindow(hwnd)
    ctypes.windll.user32.AttachThreadInput(dwCurID, dwForeID, 0)


sources/onstage_notify.py
# ClassIn Mover - A program to move ClassIn classroom window in order to
# exit from focused learning mode.

# Copyright (C) 2020-2022  Weiqi Gao, Jize Guo, Yiming Geng

# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, version 3 of the License.

# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.

# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <https://www.gnu.org/licenses/>.

import ctypes
import logging
import re
import struct
import threading
import tkinter.messagebox

import hardwareUsageDetect
import shared


def filetime2timestamp_ms(filetime):
    return filetime // 10000 - 11644473600000


class on_stage_notify:
    notify_types = {0: "None", 1: "MsgBox", 2: "MsgBox model", 3: "Switch To", 4: "Close Window", 5: "Stop Program"}

    def __init__(self, notify_type, notify_in_classroom, GetText, WindowSelector):
        self.cam = hardwareUsageDetect.WebcamDetect()
        self.mic = hardwareUsageDetect.MicDetect()
        if self.cam.not_supported or self.mic.not_supported:
            self.not_supported = True
        else:
            self.not_supported = False
        logging.debug({"on-stage notify supported": not self.not_supported})
        self.cam_used = False
        self.mic_used = False
        self.multi_window_warned = False
        self.notify_type = notify_type
        self.GetText = GetText
        self.WindowSelector = WindowSelector
        self.notify_in_classroom = notify_in_classroom

    def __getitem__(self, __name):
        return self.notify_types[__name]

    def set_notify_type(self, notify_type):
        self.notify_type = notify_type
        shared.SetSetting("on-stage notify", notify_type)

    def set_notify_in_classroom(self, notify_in_classroom):
        self.notify_in_classroom = notify_in_classroom
        shared.SetSetting("notify-in-classroom", notify_in_classroom)

    def detect_and_notify(self):
        if not self.notify_type:
            return
        cam = self.cam.getActiveApps()
        mic = self.mic.getActiveApps()
        cam_using = None
        mic_using = None
        for i in cam:
            if i.lower().endswith("classin.exe"):
                cam_using = cam[i]
        for i in mic:
            if i.lower().endswith("classin.exe"):
                mic_using = mic[i]
        notify = (cam_using and not self.cam_used) or (mic_using and not self.mic_used)
        self.cam_used = cam_using is not None
        self.mic_used = mic_using is not None
        if notify:
            if not self.notify_in_classroom:
                hwnd = ctypes.windll.user32.GetForegroundWindow()
                pid = struct.pack("l", (0))
                ctypes.windll.user32.GetWindowThreadProcessId(hwnd, pid)
                pid = struct.unpack("l", pid)[0]
                ProcessHandle = ctypes.windll.kernel32.OpenProcess(0x0410, 0, pid)
                if ProcessHandle != 0:
                    Name_Buffer = ctypes.create_unicode_buffer("", 260)
                    ctypes.windll.psapi.GetModuleFileNameExW(ProcessHandle, 0, Name_Buffer, 260)
                    ctypes.windll.kernel32.CloseHandle(ProcessHandle)
                    if Name_Buffer.value.lower().endswith("classin.exe"):
                        return
            if self.notify_type == 1:
                threading.Thread(
                    target=self._notify_msgbox, args=(cam_using is not None, mic_using is not None)
                ).start()
            elif self.notify_type == 2:
                threading.Thread(
                    target=self._notify_msgbox, args=(cam_using is not None, mic_using is not None, True)
                ).start()
            elif self.notify_type == 3:
                self._switch_to()
            elif self.notify_type == 4:
                self._close_window()
            elif self.notify_type == 5:
                self._stop_process()

    @shared.ShowThread
    def warn_multi_windows(self):
        if (not self.multi_window_warned) and (self.notify_type >= 3):
            # self.multi_window_warned = True
            tkinter.messagebox.showwarning(self.GetText("Warning"), self.GetText("multi window on-stage"))

    @shared.ShowThread
    def _notify_msgbox(self, cam=False, mic=False, model=False):
        if not (cam or mic):
            return
        if model:
            ctypes.windll.user32.MessageBoxW(
                ctypes.windll.user32.GetForegroundWindow(),
                ctypes.create_unicode_buffer(
                    (self.GetText("cam open") if cam else "") + "\n" + (self.GetText("mic open") if mic else "")
                ),
                ctypes.create_unicode_buffer("ClassIn Mover " + self.GetText("on-stage notify")),
                0x00001030,
            )
        else:
            ctypes.windll.user32.MessageBoxW(
                0,
                ctypes.create_unicode_buffer(
                    (self.GetText("cam open") if cam else "") + "\n" + (self.GetText("mic open") if mic else "")
                ),
                ctypes.create_unicode_buffer("ClassIn Mover " + self.GetText("on-stage notify")),
                0x00001030,
            )

    def _switch_to(self):
        s = self.WindowSelector.get()
        if len(s) == 0:
            return
        logging.info("Switching to %d" % int(re.findall("^\\d+", s)[0]))
        shared.SwitchWindow(int(re.findall("^\\d+", s)[0]))

    def _close_window(self):
        s = self.WindowSelector.get()
        if len(s) == 0:
            return
        ctypes.windll.user32.SendMessageW(int(re.findall("^\\d+", s)[0]), 0x10, 0, 0)

    def _stop_process(self):
        s = self.WindowSelector.get()
        if len(s) == 0:
            return
        h = ctypes.windll.kernel32.OpenProcess(0x1, 0, int(re.findall("[Pp][Ii][Dd]=(\\d+)", s)[0]))
        if not h:
            return
        ctypes.windll.kernel32.TerminateProcess(h, 1)
        ctypes.windll.kernel32.CloseHandle(h)


sources/hardwareUsageDetect.py
# Modified from https://gist.github.com/cobryan05/8e191ae63976224a0129a8c8f376adc6

import logging
import traceback
import winreg


class WebcamDetect:
    REG_KEY = winreg.HKEY_CURRENT_USER
    WEBCAM_REG_SUBKEY = (
        "SOFTWARE\\Microsoft\\Windows\\CurrentVersion\\CapabilityAccessManager\\ConsentStore\\webcam\\NonPackaged"
    )
    WEBCAM_TIMESTAMP_VALUE_NAME = "LastUsedTimeStop"
    WEBCAM_TIMESTAMP_VALUE_NAME_START = "LastUsedTimeStart"

    def __init__(self):
        self.not_supported = False
        try:
            self._regKey = winreg.OpenKey(self.REG_KEY, self.WEBCAM_REG_SUBKEY)
        except:
            logging.critical(traceback.format_exc())
            self.not_supported = True

    def getActiveApps(self):
        activeApps = {}
        # Enumerate over the subkeys of the webcam key
        subkeyCnt, valueCnt, lastModified = winreg.QueryInfoKey(self._regKey)
        for idx in range(subkeyCnt):
            subkeyName = winreg.EnumKey(self._regKey, idx)
            subkeyFullName = f"{self.WEBCAM_REG_SUBKEY}\\{subkeyName}"

            # Open each subkey and check the 'stopped' timestamp value. A value of 0 implies the camera is in use.
            subkey = winreg.OpenKey(self.REG_KEY, subkeyFullName)
            stoppedTimestamp, _ = winreg.QueryValueEx(subkey, self.WEBCAM_TIMESTAMP_VALUE_NAME)
            if 0 == stoppedTimestamp:
                activeApps[subkeyName.replace("#", "\\")] = winreg.QueryValueEx(
                    subkey, self.WEBCAM_TIMESTAMP_VALUE_NAME_START
                )[0]

        return activeApps

    def isActive(self):
        return len(self.getActiveApps()) > 0


class MicDetect:
    REG_KEY = winreg.HKEY_CURRENT_USER
    MIC_REG_SUBKEY = (
        "SOFTWARE\\Microsoft\\Windows\\CurrentVersion\\CapabilityAccessManager\\ConsentStore\\microphone\\NonPackaged"
    )
    MIC_TIMESTAMP_VALUE_NAME = "LastUsedTimeStop"
    MIC_TIMESTAMP_VALUE_NAME_START = "LastUsedTimeStart"

    def __init__(self):
        self.not_supported = False
        try:
            self._regKey = winreg.OpenKey(self.REG_KEY, self.MIC_REG_SUBKEY)
        except:
            logging.critical(traceback.format_exc())
            self.not_supported = True

    def getActiveApps(self):
        activeApps = {}
        # Enumerate over the subkeys of the microphone key
        subkeyCnt, valueCnt, lastModified = winreg.QueryInfoKey(self._regKey)
        for idx in range(subkeyCnt):
            subkeyName = winreg.EnumKey(self._regKey, idx)
            subkeyFullName = f"{self.MIC_REG_SUBKEY}\\{subkeyName}"

            # Open each subkey and check the 'stopped' timestamp value. A value of 0 implies the microphone is in use.
            subkey = winreg.OpenKey(self.REG_KEY, subkeyFullName)
            stoppedTimestamp, _ = winreg.QueryValueEx(subkey, self.MIC_TIMESTAMP_VALUE_NAME)
            if 0 == stoppedTimestamp:
                activeApps[subkeyName.replace("#", "\\")] = winreg.QueryValueEx(
                    subkey, self.MIC_TIMESTAMP_VALUE_NAME_START
                )[0]

        return activeApps

    def isActive(self):
        return len(self.getActiveApps()) > 0


Classin_Mover_Classic.py
# ClassIn Mover - A program to move ClassIn classroom window in order to
# exit from focused learning mode.

# Copyright (C) 2020-2022  Weiqi Gao, Jize Guo, Yiming Geng

# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, version 3 of the License.

# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.

# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <https://www.gnu.org/licenses/>.

__version__ = "2.0.0c"

import ctypes
import ctypes.wintypes
import datetime
import math
import os
import pathlib
import struct
import sys
import threading
import time
import tkinter
import tkinter.ttk

import psutil

import shlex4all

ClassInHwnd = []
ClassInTitle = []
ClassInPID = []
run = True
NoAdmin = False


def has_admin():
    try:
        # only windows users with admin privileges can read the C:\windows\temp
        os.listdir(os.sep.join([os.environ.get("SystemRoot", "C:\\windows"), "temp"]))
    except:
        return os.environ["USERNAME"], False
    else:
        return os.environ["USERNAME"], True


if __name__ == "__main__" and not has_admin()[1]:
    p = psutil.Process().cmdline()
    res = ctypes.windll.shell32.ShellExecuteW(
        0,
        ctypes.create_unicode_buffer("runas"),
        ctypes.create_unicode_buffer(p[0]),
        ctypes.create_unicode_buffer(shlex4all.join(p[1:])),
        0,
        5,
    )
    if res > 32:
        raise SystemExit
    NoAdmin = True


@ctypes.WINFUNCTYPE(ctypes.c_int, ctypes.wintypes.HWND, ctypes.wintypes.LPARAM)
def EnumWindowCallback(hwnd, lParam):
    global ClassInHwnd, ClassInPID, ClassInTitle
    textlen = ctypes.windll.user32.GetWindowTextLengthW(hwnd)
    text = ctypes.create_unicode_buffer("", textlen + 20)
    ctypes.windll.user32.GetWindowTextW(hwnd, text, textlen + 20)
    Caption = text.value
    if not Caption.startswith("Classroom_"):
        return 1
    pid = struct.pack("l", (0))
    ctypes.windll.user32.GetWindowThreadProcessId(hwnd, pid)
    pid = struct.unpack("l", pid)[0]
    processHandle = ctypes.windll.kernel32.OpenProcess(0x0410, 0, pid)
    if processHandle == 0:
        return 1
    name_buffer = ctypes.create_unicode_buffer("", 260)
    ctypes.windll.psapi.GetModuleFileNameExW(processHandle, 0, name_buffer, 260)
    ctypes.windll.kernel32.CloseHandle(processHandle)
    if not name_buffer.value.lower().endswith("classin.exe"):
        return 1
    ClassInHwnd.append(hwnd)
    ClassInTitle.append(Caption)
    ClassInPID.append(pid)
    return 1


def GetClassInHwnd():
    global ClassInHwnd, ClassInPID, ClassInTitle
    ClassInHwnd = []
    ClassInTitle = []
    ClassInPID = []
    ctypes.windll.user32.EnumWindows(EnumWindowCallback, 0)
    return list(zip(ClassInPID, ClassInHwnd, ClassInTitle))


def GetWindowCmd(showCmd):
    if showCmd in [2, 6]:
        return "Minimized"
    elif showCmd == 3:
        return "Maximized"
    elif showCmd == 1:
        return "Normal"
    else:
        return str(showCmd)


def PatchWindow():
    global run
    try:
        while run:
            count = 0
            st = time.time()
            CIHwnd = GetClassInHwnd()
            if len(CIHwnd) == 0:
                AddText("[%s] Cannot find classroom window\n" % datetime.datetime.now().strftime("%H:%M:%S"))
                wait = math.ceil(st) - time.time()
                time.sleep(wait if wait >= 0 else 0)
                continue
            CIHwnd = CIHwnd[0]
            while math.ceil(st) - time.time() >= 0.2 or count == 0:
                wp = struct.pack(
                    "III ll ll llll llll", *([struct.calcsize("III ll ll llll llll")] + [0] * 14)
                )  # WINDOWPLACEMENT
                ctypes.windll.user32.GetWindowPlacement(CIHwnd[1], wp)
                showCmd = struct.unpack("III ll ll llll llll", wp)[2]
                if not showCmd in [2, 3, 6]:
                    ctypes.windll.user32.ShowWindow(CIHwnd[1], 6)
                    ctypes.windll.user32.ShowWindow(CIHwnd[1], 3)
                rect = struct.pack("llll", *([0] * 4))
                ctypes.windll.user32.GetWindowRect(CIHwnd[1], rect)
                rect = struct.unpack("llll", rect)
                ctypes.windll.user32.SetWindowPos(
                    CIHwnd[1], ctypes.wintypes.HWND(-2), rect[0], rect[1], rect[2] - rect[0], rect[3] - rect[1], 3
                )
                count += 1
                time.sleep(0.05)
            AddText(
                "[%s] Classroom window is found and processed %d times:\n           PID=%d HWND=%d title=%s status=%s\n"
                % (
                    datetime.datetime.now().strftime("%H:%M:%S"),
                    count,
                    CIHwnd[0],
                    CIHwnd[1],
                    CIHwnd[2],
                    GetWindowCmd(showCmd),
                )
            )
            wait = math.ceil(st) - time.time()
            time.sleep(wait if wait >= 0 else 0)
    except:
        return


if __name__ == "__main__":
    w = tkinter.Tk()
    LogText = tkinter.Text(w, width=100, height=30, font=("Courier", 12), state=tkinter.DISABLED)
    LogText.pack(fill=tkinter.Y, side=tkinter.LEFT)
    LogScr = tkinter.ttk.Scrollbar(w, orient=tkinter.VERTICAL, command=LogText.yview)
    LogText.config(yscrollcommand=LogScr.set)
    LogScr.pack(fill=tkinter.Y, side=tkinter.RIGHT)
    w.resizable(False, False)
    w.title("ClassIn Mover Classic v2.0.0")
    w.iconbitmap(str(pathlib.Path(__file__).parent / "ClassIn_Mover.ico"))


    def AddText(text):
        LogText.config(state=tkinter.NORMAL)
        End = LogScr.get()[1] == 1
        LogText.insert(tkinter.END, text)
        if End:
            LogText.see(tkinter.END)
        LogText.config(state=tkinter.DISABLED)


    AddText("ClassIn Mover Classic v2.0.0\n")
    AddText("Copyright (C) 2020-2022  Weiqi Gao, Jize Guo\n")
    AddText("Visit https://github.com/CarlGao4/ClassIn-Mover for more information.\n\n")

    if NoAdmin:
        AddText("You are running this program without Administrator privilege, the program maybe can't function\n\n")

    PatchThread = threading.Thread(target=PatchWindow)
    PatchThread.start()

    w.mainloop()
    run = False
    sys.exit(0)
